Implement new placement strategy: 'limit-active-tasks'

aledeganopix4d opened the pull request:
This PR proposes the addition of a new placement strategy. It introduces the concept of **active task**, that is a task that is effectively running on a worker (as opposed, for instance, to build containers that can stay on the worker long after the task is finished).

A new "active_task" counter is added to the DB Worker and the responsibility of increasing/decreasing the counter is in the `task_step`.
The placement strategy considers the number of active tasks in each compatible worker and assigns **any step** to the worker with the fewest of them. Additionally a parameter "MaxActiveTasksPerWorker" can be defined, in that case workers that already have that number of active tasks **will not be selected** for tasks placement. If no worker is available with less than `MaxActiveTasksPerWorker` then the tasks will simply wait until one is free.
The worker selection in the `task_step` is serialized through a lock to prevent races where different tasks could land on the same worker.
Note that any other "step" is not restrained by `MaxActiveTasksPerWorker` and will simply chose the worker with fewest active tasks (thus `put`, `get`, `check` etc **will never be blocked**).

The PR is split in 3 commits to ease review: 
- https://github.com/concourse/concourse/commit/0a733807e1ac6cd7dd09a7edbed91de6212f0914 is the actual code only implementing the strategy and the `MaxActiveTasksPerWorker` parameter
- https://github.com/concourse/concourse/commit/887aa80357be8182788c1d5286a1aaca193a4fa8 regenerates the fakes and adapt the existing tests only
- https://github.com/concourse/concourse/commit/97138e06e34ccd015238fe55e99f6c411d2fc2a4 implements new unit tests to cover the new code

This PR supersedes #4076 and, hopefully, implements all the suggestions proposed in that discussion.

@ddadlani

kcmannem commented on atc/exec/task_step.go:
```diff
@@ -190,16 +195,57 @@ func (step *TaskStep) Run(ctx context.Context, state RunState) error {
+	var activeTasksLock lock.Lock
+
+	for {
+		if step.strategy.ModifyActiveTasks() {
```
Something minor, Can we go with with `strategy.ModifiesActiveTasks()`?

aledeganopix4d replied:
:+1:

kcmannem commented:
I like the change. Thanks for the help 👍

kcmannem commented:
idk how everyone feels about this but it might be confusing to differentiate `fewest-build-containers` and `fewest-active-tasks` at a glance. Should we go with a name like `limit-active-tasks`

aledeganopix4d commented:
> idk how everyone feels about this but it might be confusing to differentiate `fewest-build-containers` and `fewest-active-tasks` at a glance. Should we go with a name like `limit-active-tasks`

I don't have any strong opinion on the naming, just keep in mind that `MaxActiveTasksPerWorker` can be set to `0` in which case there's no limit on the active tasks, but the worker with the fewest of them is picked.
In this sense it is somewhat similar to `fewest-build-containers`.
I'll be happy to rename it to anything we're happy about.

aledeganopix4d commented:
> > idk how everyone feels about this but it might be confusing to differentiate `fewest-build-containers` and `fewest-active-tasks` at a glance. Should we go with a name like `limit-active-tasks`
> 
> I don't have any strong opinion on the naming, just keep in mind that `MaxActiveTasksPerWorker` can be set to `0` in which case there's no limit on the active tasks, but the worker with the fewest of them is picked.
> In this sense it is somewhat similar to `fewest-build-containers`.
> I'll be happy to rename it to anything we're happy about.

In any case I've followed your suggestion and the strategy is now called `limit-active-tasks`

aledeganopix4d commented on atc/engine/builder/step_factory.go:
```diff
@@ -23,6 +24,7 @@ type stepFactory struct {
 	defaultLimits         atc.ContainerLimits
 	strategy              worker.ContainerPlacementStrategy
 	resourceFactory       resource.ResourceFactory
+	lockFactory           lock.LockFactory
```
Well spotted, thanks!
I've now removed it!

aledeganopix4d commented on atc/exec/task_step.go:
```diff
@@ -190,16 +195,57 @@ func (step *TaskStep) Run(ctx context.Context, state RunState) error {
+		}
+
+		if step.strategy.ModifiesActiveTasks() {
+			if chosenWorker == nil {
```
About the first comment:
That's already been taken care of as is because `pool.allSatisfying` already returns an error when the pool is empty, so that behavior is unchanged.

About the timeout: I've thought about it. If we are to implement it should be customizable and maybe that's the right approach eventually, but for now I really wanted to keep the PR as simple as possible. We can refine the feature incrementally afterward.

xtreme-sameer-vohra replied:
Thx @aledeganopix4d

aledeganopix4d commented:
> Hey @aledeganopix4d
> Looks good. We added a few comments.
> There are some scenarios where the `decrementCounter` might not be executed (ie. the ATC is restarted ). In these scenarios, the worker would have to be `retired` and `registered` to reset the counter. It would be worth documenting this so users are aware of these edge cases.

Sure, good idea, would you know where would the best place be to document this?

xtreme-sameer-vohra commented:
> > Hey @aledeganopix4d
> > Looks good. We added a few comments.
> > There are some scenarios where the `decrementCounter` might not be executed (ie. the ATC is restarted ). In these scenarios, the worker would have to be `retired` and `registered` to reset the counter. It would be worth documenting this so users are aware of these edge cases.
> 
> Sure, good idea, would you know where would the best place be to document this?

Hey @aledeganopix4d 
We can document the `ContainerPlacementStrategy->limit-active-tasks` & `MaxActiveTasksPerWorker` flags as being `experimental` in `atc/atccmd/command.go`

And if you're so kind, you can submit a PR for `https://github.com/concourse/docs/blob/master/lit/docs/operation/container-placement.lit`. Here you can leverage the `\warn` tag to make a note that this is experimental

Thanks :)

aledeganopix4d commented:
> > > Hey @aledeganopix4d
> > > Looks good. We added a few comments.
> > > There are some scenarios where the `decrementCounter` might not be executed (ie. the ATC is restarted ). In these scenarios, the worker would have to be `retired` and `registered` to reset the counter. It would be worth documenting this so users are aware of these edge cases.
> > 
> > 
> > Sure, good idea, would you know where would the best place be to document this?
> 
> Hey @aledeganopix4d
> We can document the `ContainerPlacementStrategy->limit-active-tasks` & `MaxActiveTasksPerWorker` flags as being `experimental` in `atc/atccmd/command.go`
> 
> And if you're so kind, you can submit a PR for `https://github.com/concourse/docs/blob/master/lit/docs/operation/container-placement.lit`. Here you can leverage the `\warn` tag to make a note that this is experimental
> 
> Thanks :)

Hello @xtreme-sameer-vohra,
I've added the Experimental notice in the command line description and then opened a PR in the docs describing this new placement strategy here: https://github.com/concourse/docs/pull/231.
Thanks.

gerhard commented:
Really looking forward to v5.5 shipping with this feature. We've just hit a new wall of failed tests due to CPU contention, even though we have massively over-provisioned workers. It's either baby-sitting our pipelines so that builds go through, or waiting days for builds to go through. 🚢 :shipit:
